File: lib/coffee_script/scope.rb
module CoffeeScript

  # Scope objects form a tree corresponding to the shape of the function
  # definitions present in the script. They provide lexical scope, to determine
  # whether a variable has been seen before or if it needs to be declared.
  class Scope

    attr_reader :parent, :expressions, :function, :variables, :temp_variable

    # Initialize a scope with its parent, for lookups up the chain,
    # as well as the Expressions body where it should declare its variables,
    # and the function that it wraps.
    def initialize(parent, expressions, function)
      @parent, @expressions, @function = parent, expressions, function
      @variables = {}
      @temp_variable = @parent ? @parent.temp_variable.dup : '__a'
    end

    # Look up a variable in lexical scope, or declare it if not found.
    def find(name, remote=false)
      found = check(name)
      return found if found || remote
      @variables[name.to_sym] = :var
      found
    end

    # Define a local variable as originating from a parameter in current scope
    # -- no var required.
    def parameter(name)
      @variables[name.to_sym] = :param
    end

    # Just check to see if a variable has already been declared.
    def check(name)
      return true if @variables[name.to_sym]
      !!(@parent && @parent.check(name))
    end

    # You can reset a found variable on the immediate scope.
    def reset(name)
      @variables[name.to_sym] = false
    end

    # Find an available, short, name for a compiler-generated variable.
    def free_variable
      @temp_variable.succ! while check(@temp_variable)
      @variables[@temp_variable.to_sym] = :var
      Value.new(@temp_variable.dup)
    end

    # Ensure that an assignment is made at the top of scope (or top-level
    # scope, if requested).
    def assign(name, value, top=false)
      return @parent.assign(name, value, top) if top && @parent
      @variables[name.to_sym] = Value.new(value)
    end

    # Does this scope reference any variables that need to be declared in the
    # given function body?
    def declarations?(body)
      !declared_variables.empty? && body == @expressions
    end

    # Does this scope reference any assignments that need to be declared at the
    # top of the given function body?
    def assignments?(body)
      !assigned_variables.empty? && body == @expressions
    end

    # Return the list of variables first declared in current scope.
    def declared_variables
      @variables.select {|k, v| v == :var }.map {|pair| pair[0].to_s }.sort
    end

    # Return the list of variables that are supposed to be assigned at the top
    # of scope.
    def assigned_variables
      @variables.select {|k, v| v.is_a?(Value) }.sort_by {|pair| pair[0].to_s }
    end

    def compiled_declarations
      declared_variables.join(', ')
    end

    def compiled_assignments
      assigned_variables.map {|name, val| "#{name} = #{val}"}.join(', ')
    end

    def inspect
      "<Scope:#{__id__} #{@variables.inspect}>"
    end

  end

end

File: lib/coffee_script/scope.js
(function(){
  var Scope, succ;
  var __hasProp = Object.prototype.hasOwnProperty;
  if (!((typeof process !== "undefined" && process !== null))) {
    this.exports = this;
  }
  // Scope objects form a tree corresponding to the shape of the function
  // definitions present in the script. They provide lexical scope, to determine
  // whether a variable has been seen before or if it needs to be declared.
  //
  // Initialize a scope with its parent, for lookups up the chain,
  // as well as the Expressions body where it should declare its variables,
  // and the function that it wraps.
  Scope = (exports.Scope = function Scope(parent, expressions, method) {
    this.parent = parent;
    this.expressions = expressions;
    this.method = method;
    this.variables = {};
    this.temp_variable = this.parent ? this.parent.temp_variable : '__a';
    return this;
  });
  // Look up a variable in lexical scope, or declare it if not found.
  Scope.prototype.find = function find(name, remote) {
    var found;
    found = this.check(name);
    if (found || remote) {
      return found;
    }
    this.variables[name] = 'var';
    return found;
  };
  // Define a local variable as originating from a parameter in current scope
  // -- no var required.
  Scope.prototype.parameter = function parameter(name) {
    return this.variables[name] = 'param';
  };
  // Just check to see if a variable has already been declared.
  Scope.prototype.check = function check(name) {
    if (this.variables[name]) {
      return true;
    }
    return !!(this.parent && this.parent.check(name));
  };
  // You can reset a found variable on the immediate scope.
  Scope.prototype.reset = function reset(name) {
    return delete this.variables[name];
  };
  // Find an available, short, name for a compiler-generated variable.
  Scope.prototype.free_variable = function free_variable() {
    while (this.check(this.temp_variable)) {
      ((this.temp_variable = succ(this.temp_variable)));
    }
    this.variables[this.temp_variable] = 'var';
    return this.temp_variable;
  };
  // Ensure that an assignment is made at the top of scope (or top-level
  // scope, if requested).
  Scope.prototype.assign = function assign(name, value, top_level) {
    if (top_level && this.parent) {
      return this.parent.assign(name, value, top_level);
    }
    return this.variables[name] = {
      value: value,
      assigned: true
    };
  };
  // Does this scope reference any variables that need to be declared in the
  // given function body?
  Scope.prototype.has_declarations = function has_declarations(body) {
    return body === this.expressions && this.declared_variables().length;
  };
  // Does this scope reference any assignments that need to be declared at the
  // top of the given function body?
  Scope.prototype.has_assignments = function has_assignments(body) {
    return body === this.expressions && this.assigned_variables().length;
  };
  // Return the list of variables first declared in current scope.
  Scope.prototype.declared_variables = function declared_variables() {
    var __a, __b, key, val;
    return ((function() {
      __a = []; __b = this.variables;
      for (key in __b) if (__hasProp.call(__b, key)) {
        val = __b[key];
        if (val === 'var') {
          __a.push(key);
        }
      }
      return __a;
    }).call(this)).sort();
  };
  // Return the list of variables that are supposed to be assigned at the top
  // of scope.
  Scope.prototype.assigned_variables = function assigned_variables() {
    var __a, __b, key, val;
    return ((function() {
      __a = []; __b = this.variables;
      for (key in __b) if (__hasProp.call(__b, key)) {
        val = __b[key];
        if (val.assigned) {
          __a.push([key, val.value]);
        }
      }
      return __a;
    }).call(this)).sort();
  };
  Scope.prototype.compiled_declarations = function compiled_declarations() {
    return this.declared_variables().join(', ');
  };
  Scope.prototype.compiled_assignments = function compiled_assignments() {
    var __a, __b, __c, t;
    return ((function() {
      __a = []; __b = this.assigned_variables();
      for (__c = 0; __c < __b.length; __c++) {
        t = __b[__c];
        __a.push(t[0] + ' = ' + t[1]);
      }
      return __a;
    }).call(this)).join(', ');
  };
  // Helper functions:
  // The next character alphabetically, to produce the following string.
  succ = function succ(str) {
    return str.slice(0, str.length - 1) + String.fromCharCode(str.charCodeAt(str.length - 1) + 1);
  };
})();